src/OrchardCore/OrchardCore.Queries.Core/Services/DefaultQueryManager.cs

```diff
@@ -0,0 +1,278 @@
+        if (data != null)
+        {
+            UpdateBaseProperties(data, query);
+
+            // During the update, we allow changing the source.
+            var sourceName = data[nameof(Query.Source)]?.GetValue<string>();
+
+            if (!string.IsNullOrEmpty(sourceName) && query.Source != sourceName)
+            {
+                var querySource = _serviceProvider.GetKeyedService<IQuerySource>(sourceName);
+
+                if (querySource != null)
+                {
+                    query.Source = querySource.Name;
+                }
+            }
+        }
+
+        var updatingContext = new UpdatingQueryContext(query, data);
+        await _queryHandlers.InvokeAsync((handler, context) => handler.UpdatingAsync(context), updatingContext, _logger);
+
+        var document = await _documentManager.GetOrCreateMutableAsync();
+        document.Queries[query.Name] = query;
+        await _documentManager.UpdateAsync(document);
+
+        var updatedContext = new UpdatedQueryContext(query);
+        await _queryHandlers.InvokeAsync((handler, context) => handler.UpdatedAsync(context), updatedContext, _logger);
+    }
+
+    public async Task SaveAsync(params Query[] queries)
+    {
+        if (queries?.Length == 0)
+        {
+            return;
+        }
+
+        var document = await _documentManager.GetOrCreateMutableAsync();
+
+        foreach (var query in queries)
+        {
+            document.Queries[query.Name] = query;
+        }
+
+        await _documentManager.UpdateAsync(document);
+    }
+
+    private async Task<IEnumerable<Query>> LocateQueriesAsync(QueryContext context)
+    {
+        var document = await _documentManager.GetOrCreateImmutableAsync();
+
+        if (context == null)
+        {
+            return document.Queries.Values;
+        }
+
+        var queries = document.Queries.Values.AsEnumerable();
+
+        if (!string.IsNullOrEmpty(context.Source))
+        {
+            queries = queries.Where(x => x.Source.Equals(context.Source, StringComparison.OrdinalIgnoreCase));
+        }
+
+        if (!string.IsNullOrEmpty(context.Name))
+        {
+            queries = queries.Where(x => x.Name.Contains(context.Name, StringComparison.OrdinalIgnoreCase));
+        }
+
+        if (context.Sorted)
+        {
+            queries = queries.OrderBy(x => x.Name);
+        }
+
+        return queries;
+    }
+
+    private Task LoadAsync(Query query)
+    {
+        var loadedContext = new LoadedQueryContext(query);
+
+        return _queryHandlers.InvokeAsync((handler, context) => handler.LoadedAsync(context), loadedContext, _logger);
+    }
+
+    private static void UpdateBaseProperties(JsonNode data, Query query)
+    {
+        query.Schema = data[nameof(Query.Schema)]?.GetValue<string>();
+
+        // For backward compatibility, we use the key 'ReturnDocuments'.
+        var returnDocuments = data["ReturnDocuments"];
```

Comment:
Why not update this in `QueriesDocumentMigrationHelper` too and get rid of it once and for all?

Reply:
Because some people maybe using the old value in recipes. This way to don't break them.

Reply:
Ah OK.